vllm/v1/offloading/worker/cpu_gpu.py

```diff
@@ -0,0 +1,169 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import torch
+
+from vllm import _custom_ops as ops
+from vllm.attention import AttentionBackend
+from vllm.logger import init_logger
+from vllm.utils import is_pin_memory_available
+from vllm.v1.offloading.mediums import CPULoadStoreSpec, GPULoadStoreSpec
+from vllm.v1.offloading.worker.worker import (OffloadingHandler,
+                                              TransferResult, TransferSpec)
+
+logger = init_logger(__name__)
+
+
+def expand_block_ids(block_ids: torch.Tensor,
+                     block_size_factor: int,
+                     output: torch.Tensor,
+                     skip_count: int = 0):
+    """
+    Convert a list of block IDs to a list of matching block ids,
+    assuming each block is composed of actual block_size_factor blocks.
+    Outputs to output tensor.
+    The first skip_count blocks will be skipped.
+    Note that skip_count must be less than block_size_factor.
+
+    For example, if block_ids = [0, 1, 3] and block_size_factor =  4,
+    then it yields [0, 1, 2, 3, 4, 5, 6, 7, 12, 13, 14, 15]
+    since 0 maps to [0, 1, 2, 3]
+    1 maps to [4, 5, 6, 7]
+    and 3 maps to [12, 13, 14, 15]
+    """
+    assert skip_count < block_size_factor
+
+    output_idx = 0
+    for block_id in block_ids:
+        base_block_id = block_id * block_size_factor
+        for i in range(skip_count, block_size_factor):
+            output[output_idx] = base_block_id + i
+            output_idx += 1
+
+        # finished skipping
+        skip_count = 0
```

Comment:
updating tensor element-wise is actually very slow, I was thinking more like what you had before but passing `out` arg to `torch.cat`.

Alternatively could keep as numpy array (which I think block_ids should be already per other comment) ... which are much more efficient to manipulate.

By the way I don't think these specific optimizations should hold up getting the functionality in, can always be follow-on updates.